subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/InstrumentingClassTransform.java

```diff
@@ -188,29 +104,59 @@ public InstrumentingClassTransform(BytecodeInterceptorFilter interceptorFilter)
         this.externalInterceptors = CallInterceptorRegistry.getJvmBytecodeInterceptors(interceptorFilter);
     }
 
+    private BytecodeInterceptorFilter interceptorFilter() {
+        return externalInterceptors.getOriginalFilter();
+    }
+
+    private List<JvmBytecodeCallInterceptor> buildInterceptors(InstrumentationMetadata metadata) {
+        return externalInterceptors.getInterceptors(metadata);
+    }
+
     @Override
     public Pair<RelativePath, ClassVisitor> apply(ClasspathEntryVisitor.Entry entry, ClassVisitor visitor, ClassData classData) {
+        // TODO(mlopatkin) can we reuse interceptors in a bigger scope, not per class, but per artifact?
+        List<JvmBytecodeCallInterceptor> interceptors = buildInterceptors(classData);
+        if (interceptorFilter().matches(ADHOC_INTERCEPTORS)) {
+            interceptors = ImmutableList.<JvmBytecodeCallInterceptor>builderWithExpectedSize(interceptors.size() + 1).add(ADHOC_INTERCEPTORS).addAll(interceptors).build();
+        }
         return Pair.of(entry.getPath(),
             new InstrumentingVisitor(
                 new CallInterceptionClosureInstrumentingClassVisitor(
                     new LambdaSerializationTransformer(new InstrumentingBackwardsCompatibilityVisitor(visitor)),
-                    externalInterceptors.getOriginalFilter()
+                    interceptorFilter()
                 ),
-                classData, externalInterceptors
+                classData,
+                interceptors,
+                interceptorFilter()
             )
         );
     }
 
+    private static class BridgeMethod {
+        final Handle bridgeMethodHandle;
+        final BridgeMethodBuilder bridgeMethodBuilder;
+
+        private BridgeMethod(Handle bridgeMethodHandle, BridgeMethodBuilder bridgeMethodBuilder) {
+            this.bridgeMethodHandle = bridgeMethodHandle;
+            this.bridgeMethodBuilder = bridgeMethodBuilder;
+        }
+    }
+
     private static class InstrumentingVisitor extends ClassVisitor {
-        String className;
         private final ClassData classData;
+        private final List<JvmBytecodeCallInterceptor> interceptors;
+        private final BytecodeInterceptorFilter interceptorFilter;
+
+        final Map<Handle, BridgeMethod> bridgeMethods = new HashMap<>();
+
+        private String className;
         private boolean hasGroovyCallSites;
-        private final JvmBytecodeInterceptorSet externalInterceptors;
 
-        public InstrumentingVisitor(ClassVisitor visitor, ClassData classData, JvmBytecodeInterceptorSet externalInterceptors) {
+        public InstrumentingVisitor(ClassVisitor visitor, ClassData classData, List<JvmBytecodeCallInterceptor> interceptors, BytecodeInterceptorFilter interceptorFilter) {
```

Comment:
💭 This is mostly a drive-by improvement: here we save a bit on allocating all these interceptors per class rather than per method.